File: tools/python/validate/validate_all.py
#!/usr/bin/env python3
"""Validate AOM examples against their schemas. Optionally restrict to one folder."""

import glob
import subprocess
import sys
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parent.parent.parent.parent
VALIDATOR = Path(__file__).resolve().parent / "validate.py"
SCHEMAS = REPO_ROOT / "spec" / "v0.1.0"


def _examples_root():
    for name in ("Examples", "examples"):
        d = REPO_ROOT / name
        if d.is_dir():
            return d
    return REPO_ROOT / "Examples"


def _run_validation(schema_path: Path, files: list[str], kind: str, failures: list[dict]):
    """Run the generic validator for each file, but do not stop on first failure.

    Accumulates failures in `failures` as dicts: {"kind", "rel", "details"}.
    """
    for f in files:
        cmd = f'python "{VALIDATOR}" "{schema_path}" "{f}"'
        result = subprocess.run(cmd, shell=True, cwd=REPO_ROOT, capture_output=True, text=True)
        rel = Path(f).resolve().relative_to(REPO_ROOT).as_posix()
        if result.returncode != 0:
            print(f"[FAIL] {rel}")
            details = (result.stdout or "") + (result.stderr or "")
            failures.append({"kind": kind, "rel": rel, "details": details})
        else:
            print(f"[OK] {rel}")


def main():
    examples_root = _examples_root()
    if len(sys.argv) >= 2:
        folder_arg = sys.argv[1].strip().replace("\\", "/")
        if Path(folder_arg).is_absolute():
            search_root = Path(folder_arg)
        else:
            candidates = [
                (examples_root / folder_arg).resolve(),
                (REPO_ROOT / folder_arg).resolve(),
            ]
            if "/" in folder_arg and folder_arg.lower().startswith("examples/"):
                rest = folder_arg.split("/", 1)[1]
                candidates.append((REPO_ROOT / examples_root.name / rest).resolve())
            search_root = None
            for c in candidates:
                if c.is_dir():
                    search_root = c
                    break
            if search_root is None:
                search_root = candidates[0]
        if not search_root.is_dir():
            print(f"[FAIL] Folder not found: {search_root}")
            sys.exit(1)
        print(f"Validating only: {search_root.relative_to(REPO_ROOT)}\n")
    else:
        search_root = examples_root
        print(f"Validating all examples under: {search_root.relative_to(REPO_ROOT)}\n")

    aom_glob = str(search_root / "**" / "*.aom.json")
    out_glob = str(search_root / "**" / "*.output.json")
    skip_glob = str(search_root / "**" / "*.output.json.skip")
    aom_files = sorted(glob.glob(aom_glob, recursive=True))
    output_files = sorted(glob.glob(out_glob, recursive=True))
    skip_files = sorted(glob.glob(skip_glob, recursive=True))
    output_files = output_files + skip_files

    print(f"Found {len(aom_files)} AOM surface(s), {len(output_files)} output(s)\n")

    core_schema = SCHEMAS / "aom-input-schema.json"
    output_schema = SCHEMAS / "aom-output-schema.json"
    if not core_schema.is_file():
        print("Schema not found:", core_schema)
        sys.exit(1)

    failures: list[dict] = []

    print("aom-input-schema.json")
    _run_validation(core_schema, aom_files, "input", failures)

    print("\naom-output-schema.json")
    _run_validation(output_schema, output_files, "output", failures)

    if failures:
        print("\nSummary of failed files:")
        input_failures = [f for f in failures if f["kind"] == "input"]
        output_failures = [f for f in failures if f["kind"] == "output"]

        if input_failures:
            print("  Input surfaces (aom.json):")
            for f in input_failures:
                print(f"    - {f['rel']}")
                details = (f.get("details") or "").strip()
                if details:
                    print("      details:")
                    for line in details.splitlines():
                        print(f"        {line}")

        if output_failures:
            print("  Outputs (output.json / output.json.skip):")
            for f in output_failures:
                print(f"    - {f['rel']}")
                details = (f.get("details") or "").strip()
                if details:
                    print("      details:")
                    for line in details.splitlines():
                        print(f"        {line}")

        sys.exit(1)

    print("\nAll valid!")


if __name__ == "__main__":
    main()

File: tools/node/validate/validate_all.js
#!/usr/bin/env node
/**
 * Validate AOM examples against their schemas. Optionally restrict to one folder.
 * Usage: node validate_all.js [folder]
 */

import { spawnSync } from "child_process";
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));

const REPO_ROOT = path.resolve(__dirname, "..", "..", "..");
const VALIDATOR = path.join(__dirname, "validate.js");

function getSchemasDir() {
  const specV = path.join(REPO_ROOT, "spec", "v0.1.0");
  if (fs.existsSync(specV) && fs.statSync(specV).isDirectory()) return specV;
  return specV;
}

function examplesRoot() {
  for (const name of ["Examples", "examples"]) {
    const d = path.join(REPO_ROOT, name);
    if (fs.existsSync(d) && fs.statSync(d).isDirectory()) return d;
  }
  return path.join(REPO_ROOT, "Examples");
}

function globRecursive(dir, suffix) {
  const results = [];
  function walk(d) {
    if (!fs.existsSync(d) || !fs.statSync(d).isDirectory()) return;
    for (const name of fs.readdirSync(d)) {
      const full = path.join(d, name);
      const stat = fs.statSync(full);
      if (stat.isDirectory()) walk(full);
      else if (stat.isFile() && full.endsWith(suffix)) results.push(full);
    }
  }
  walk(dir);
  return results.sort();
}

function runValidator(schemaPath, jsonPath) {
  const result = spawnSync(process.execPath, [VALIDATOR, schemaPath, jsonPath], {
    cwd: REPO_ROOT,
    encoding: "utf-8",
    stdio: ["ignore", "pipe", "pipe"],
  });
  return { status: result.status, stderr: result.stderr || "" };
}

function main() {
  const examplesRootDir = examplesRoot();
  let searchRoot;

  if (process.argv.length >= 3) {
    const folderArg = process.argv[2].trim().replace(/\\/g, "/");
    if (path.isAbsolute(folderArg)) {
      searchRoot = path.resolve(folderArg);
    } else {
      const candidates = [
        path.resolve(examplesRootDir, folderArg),
        path.resolve(REPO_ROOT, folderArg),
      ];
      if (folderArg.includes("/") && folderArg.toLowerCase().startsWith("examples/")) {
        const rest = folderArg.split("/").slice(1).join(path.sep);
        candidates.push(path.resolve(REPO_ROOT, path.basename(examplesRootDir), rest));
      }
      searchRoot = null;
      for (const c of candidates) {
        if (fs.existsSync(c) && fs.statSync(c).isDirectory()) {
          searchRoot = c;
          break;
        }
      }
      if (searchRoot === null) searchRoot = candidates[0];
    }
    if (!fs.existsSync(searchRoot) || !fs.statSync(searchRoot).isDirectory()) {
      console.error("Folder not found: " + searchRoot);
      process.exit(1);
    }
    console.log("Validating only: " + path.relative(REPO_ROOT, searchRoot) + "\n");
  } else {
    searchRoot = examplesRootDir;
    console.log("Validating all examples under: " + path.relative(REPO_ROOT, searchRoot) + "\n");
  }

  const aomFiles = globRecursive(searchRoot, ".aom.json");
  const outputFiles = globRecursive(searchRoot, ".output.json");
  const skipFiles = globRecursive(searchRoot, ".output.json.skip");

  console.log("Found " + aomFiles.length + " AOM surface(s), " + (outputFiles.length + skipFiles.length) + " output(s)\n");

  const SCHEMAS = getSchemasDir();
  const coreSchema = path.join(SCHEMAS, "aom-input-schema.json");
  const outputSchema = path.join(SCHEMAS, "aom-output-schema.json");
  if (!fs.existsSync(coreSchema) || !fs.statSync(coreSchema).isFile()) {
    console.error("Schema not found: " + coreSchema);
    process.exit(1);
  }

  const failures = [];

  console.log("aom-input-schema.json");
  for (const f of aomFiles) {
    const rel = path.relative(REPO_ROOT, f).replace(/\\/g, "/");
    const { status, stderr } = runValidator(coreSchema, f);
    if (status !== 0) {
      console.log("[FAIL] " + rel);
      failures.push({ kind: "input", rel, stderr });
    } else {
      console.log("[OK] " + rel);
    }
  }

  console.log("\naom-output-schema.json");
  for (const f of [...outputFiles, ...skipFiles]) {
    const rel = path.relative(REPO_ROOT, f).replace(/\\/g, "/");
    const { status, stderr } = runValidator(outputSchema, f);
    if (status !== 0) {
      console.log("[FAIL] " + rel);
      failures.push({ kind: "output", rel, stderr });
    } else {
      console.log("[OK] " + rel);
    }
  }

  if (failures.length) {
    console.log("\nSummary of failed files:");
    const inputFailures = failures.filter((f) => f.kind === "input");
    const outputFailures = failures.filter((f) => f.kind === "output");

    if (inputFailures.length) {
      console.log("  Input surfaces (aom.json):");
      for (const { rel, stderr } of inputFailures) {
        console.log("    - " + rel);
        if (stderr) {
          console.log("      details:");
          stderr
            .trim()
            .split("\n")
            .forEach((line) => console.log("        " + line));
        }
      }
    }

    if (outputFailures.length) {
      console.log("  Outputs (output.json / output.json.skip):");
      for (const { rel, stderr } of outputFailures) {
        console.log("    - " + rel);
        if (stderr) {
          console.log("      details:");
          stderr
            .trim()
            .split("\n")
            .forEach((line) => console.log("        " + line));
        }
      }
    }

    process.exit(1);
  }

  console.log("\nAll valid!");
}

main();
